### ISSUE TITLE
Hold multiple synapse versions in apt repository

### ISSUE THREAD

**Issue by a-0-dev** [matrix-org/synapse#14082]
Currently, only the latest ´matrix-synapse-py3` version is available in the apt repository, which poses challenges when managing synapse with ansible or other tools which may require to install/retain an older version if updates are not yet desired.

Leaving older versions in the repository (as it is the case for most software packages) would significantly support these admins :)

This may also be classified as a bug, depending on your criteria.

**Comment by domrim** [comment]
this is unfortunately not very usable, as they are not in the [package index](https://packages.matrix.org/debian/dists/bullseye/main/binary-amd64/Packages). There are only 1.77.0+bullseye1 and 1.71.0+bullseye1 available (not sure why exactly these two versions), would it be possible to keep at least the last 3-5 versions in the package index?

**Comment by richvdh** [comment]
> this is unfortunately not very usable, as they are not in the [package index](https://packages.matrix.org/debian/dists/bullseye/main/binary-amd64/Packages).

No, but downloading them and installing them with `dpkg` is not exactly the most onerous task in the world. Indeed, I'm not sure it's any harder than figuring out the incantation to get `apt` to install the version you want.

It looks like support for multiple versions of packages was added in a recent version of `reprepro` (see https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=570623), so this might be something we *could* consider now. I suspect the thing that makes 1.71.0 different is that the 1.72.0 release was done by someone using `reprepro` 5.4.0. (@erikjohnston  looks like it was you: can you confirm?)

The problem is that doing so would require all of the Synapse core team to be using bleeding-edge reprepro.

**Comment by erikjohnston** [comment]
> (@erikjohnston looks like it was you: can you confirm?)

My local `reprepro` appears to be `5.3.0`. It's possible someone in the office did the debs that time round?

**Comment by domrim** [comment]
> No, but downloading them and installing them with `dpkg` is not exactly the most onerous task in the world. Indeed, I'm not sure it's any harder than figuring out the incantation to get `apt` to install the version you want.

Yes and no. The advantage of `apt` is, that it checks the signatures of the repository and package directly. One could argue that in a larger environment with multiple servers this could get automated, but it's not trivial and has some pitfalls.
`apt` usage with a version on the other hand is trivial with `apt install matrix-synapse-py3=1.76.0+bullseye1`

**Comment by richvdh** [comment]
I'll reopen this for the team to consider
